### PR TITLE
rbd-mirror: continue to isolate journal replay logic

### PR DESCRIPTION
**Pull request by dillaman**
There will be one additional PR after this one to remove all remaining traces of journal-only logic and variables from the shared image replaying state machines.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dillaman** [comment]
http://pulpito.ceph.com/jdillaman-2019-12-22_21:09:10-rbd-wip-jd-testing-distro-basic-smithi/

File: src/tools/rbd_mirror/ImageReplayer.cc [hunk 1]
```diff
@@ -961,25 +961,27 @@ template <typename I>
...
+    std::shared_lock image_locker{m_local_image_ctx->image_lock};
+    if (m_local_image_name != m_local_image_ctx->name) {
       // will re-register with new name after next status update
       dout(10) << "image renamed" << dendl;
-      m_local_image_name = local_image_name;
+      m_local_image_name = m_local_image_ctx->name;
     }
   }
 
+  // replayer cannot shut down until after callback completes
```
Comment (trociny) [inline src/tools/rbd_mirror/ImageReplayer.cc]: @dillaman Was it intentional to add this comment and leave the one below? They seem to me very similar?

Comment (dillaman) [reply src/tools/rbd_mirror/ImageReplayer.cc]: Missed the second one

File: src/tools/rbd_mirror/image_replayer/BootstrapRequest.cc [hunk 2]
```diff
@@ -607,16 +363,11 @@ void BootstrapRequest<I>::handle_image_sync(int r) {
...
-    if (m_canceled) {
-      dout(10) << "request canceled" << dendl;
-      m_ret_val = -ECANCELED;
-    }
-
-    if (r < 0) {
-      derr << "failed to sync remote image: " << cpp_strerror(r) << dendl;
-      m_ret_val = r;
-    }
+  if (r < 0) {
```
Comment (trociny) [inline src/tools/rbd_mirror/image_replayer/BootstrapRequest.cc]: nit: I think the check for `m_canceled` protected from a potential race when `m_image_sync->cancel()` was called too late to make the image sync request return the "canceled" error code. It looks like it does not hurt though and your change is ok.

But it seems like we unintentionally (in the original version and in yours) output the error message for the case when the image sync request is cancelled. May be fix this while here?

**Comment by trociny** [comment]
@dillaman observing this failuer [1], which I think actually not related to this PR but to the previous one. It fails on "delete images during bootstrap" stress test, waiting for the secondary image is being deleted after mirroring is disabled. In the rbd-mirror logs we can see on the disable:
```
/ceph/teuthology-archive/trociny-2019-12-24_14:17:10-rbd-wip-mgolub-testing-distro-basic-smithi/4629894/remote/smithi096/log/cluster1-client.mirror.1.15015.log.gz:

2019-12-24T14:56:03.440+0000 7eff7a7b4700 10 rbd::mirror::ImageReplayer: 0x55f3e7d2cd00 [2/3715ee20-68a1-4f34-b42d-6300474f3fcf] stop: on_finish=0x55f3eb59d4e0, manual=0, desc=
2019-12-24T14:56:03.440+0000 7eff7a7b4700 10 rbd::mirror::ImageReplayer: 0x55f3e7d2cd00 [2/3715ee20-68a1-4f34-b42d-6300474f3fcf] stop: canceling start
2019-12-24T14:56:03.440+0000 7eff7a7b4700 10 rbd::mirror::InstanceWatcher: 0x55f3e95e4000 complete_request: instance_id=4299, request_id=263
2019-12-24T14:56:03.444+0000 7eff7a7b4700 10 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 handle_init_remote_journaler: r=-2
2019-12-24T14:56:03.444+0000 7eff7a7b4700 -1 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 handle_init_remote_journaler: failed to initialize remote journal: (2) No such file or directory
2019-12-24T14:56:03.444+0000 7eff7a7b4700 10 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 handle_replay_complete: r=-2, error=error initializing remote journal
2019-12-24T14:56:03.444+0000 7eff7a7b4700 -1 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 handle_replay_complete: replay encountered an error: (2) No such file or directory
2019-12-24T14:56:03.444+0000 7eff7a7b4700 10 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 close_local_image: 
2019-12-24T14:56:03.444+0000 7eff7a7b4700 20 rbd::mirror::image_replayer::CloseImageRequest: 0x55f3eafe7d00 close_image
2019-12-24T14:56:03.444+0000 7eff6678c700 20 rbd::mirror::MirrorStatusWatcher: 0x55f3e7070940 handle_notify: 
2019-12-24T14:56:03.452+0000 7eff53766700 20 rbd::mirror::image_replayer::CloseImageRequest: 0x55f3eafe7d00 handle_close_image: r=0
2019-12-24T14:56:03.452+0000 7eff53766700 10 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 handle_close_local_image: r=0
2019-12-24T14:56:03.452+0000 7eff53766700 10 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 shut_down_remote_journaler: 
2019-12-24T14:56:03.452+0000 7eff7a7b4700 10 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 handle_shut_down_remote_journaler: r=0
2019-12-24T14:56:03.452+0000 7eff7a7b4700 10 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 wait_for_in_flight_ops: 
2019-12-24T14:56:03.452+0000 7eff7a7b4700 10 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 handle_wait_for_in_flight_ops: r=0
2019-12-24T14:56:03.452+0000 7eff7a7b4700 10 rbd::mirror::ImageReplayer: 0x55f3e7d2cd00 [2/3715ee20-68a1-4f34-b42d-6300474f3fcf] handle_start_replay: r=-2
2019-12-24T14:56:03.452+0000 7eff7a7b4700 10 rbd::mirror::ImageReplayer: 0x55f3e7d2cd00 [2/3715ee20-68a1-4f34-b42d-6300474f3fcf] on_start_fail: r=-125, desc=
2019-12-24T14:56:03.452+0000 7eff7a7b4700 10 rbd::mirror::ImageReplayer: 0x55f3e7d2cd00 [2/3715ee20-68a1-4f34-b42d-6300474f3fcf] operator(): start canceled
2019-12-24T14:56:03.452+0000 7eff7a7b4700 10 rbd::mirror::ImageReplayer: 0x55f3e7d2cd00 [2/3715ee20-68a1-4f34-b42d-6300474f3fcf] set_state_description: r=-125, desc=
2019-12-24T14:56:03.452+0000 7eff7a7b4700 15 rbd::mirror::ImageReplayer: 0x55f3e7d2cd00 [2/3715ee20-68a1-4f34-b42d-6300474f3fcf] update_mirror_image_status: force=0, state=--
2019-12-24T14:56:03.452+0000 7eff7a7b4700 15 rbd::mirror::ImageReplayer: 0x55f3e7d2cd00 [2/3715ee20-68a1-4f34-b42d-6300474f3fcf] update_mirror_image_status: shut down in-progress: ignoring update
2019-12-24T14:56:03.452+0000 7eff7a7b4700 10 rbd::mirror::ImageReplayer: 0x55f3e7d2cd00 [2/3715ee20-68a1-4f34-b42d-6300474f3fcf] shut_down: r=-125
2019-12-24T14:56:03.452+0000 7eff7a7b4700 10 rbd::mirror::image_replayer::journal::Replayer: 0x55f3eb8f9000 shut_down: 
```
That  is when it calls the replayer shutdown, I think we are in "raced with the last piece of the init state machine" condition [2], and in this branch I think we are missing `m_on_init_shutdown` context completion before returning.

[1] http://qa-proxy.ceph.com/teuthology/trociny-2019-12-24_14:17:10-rbd-wip-mgolub-testing-distro-basic-smithi/4629894/teuthology.log
[2] https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/image_replayer/journal/Replayer.cc#L229

File: src/tools/rbd_mirror/image_replayer/BootstrapRequest.cc (hunk 2 above)
Comment (dillaman) [reply src/tools/rbd_mirror/image_replayer/BootstrapRequest.cc]: No longer log it via `derr`

**Comment by dillaman** [comment]
jenkins test make check
(unrelated failure)

**Comment by dillaman** [comment]
jenkins test make check

**Pull request merged**